`sin`: port to structured

hameerabbasi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52277 Port sin to structured.**
* #52276 Remove hacky wrapper from a lot of unary operators.

Differential Revision: [D26732398](https://our.internmc.facebook.com/intern/diff/D26732398)

hameerabbasi committed:
Port sinc to structured.

[ghstack-poisoned]

hameerabbasi committed:
Update on "Port sinc to structured."




[ghstack-poisoned]

hameerabbasi commented:
cc @bdhirsh Would you happen to know what's going on here?

kshitij12345 commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -23,6 +23,17 @@
+TORCH_META_FUNC(sin) (
+  const Tensor& self, IntArrayRef output_size
+) {
+  build(unary_float_op(maybe_get_output(), self));
```
This should be `TensorIterator::unary_float_op` I think. Could you try that?

kshitij12345 commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -346,9 +357,10 @@ Tensor& sgn_out(const Tensor& self, Tensor& result) {
-Tensor& sin_(Tensor& self) { return unary_op_impl_(self, at::sin_out); }
+TORCH_IMPL_FUNC(sin_out) (const Tensor& self, const Tensor& result) {
+  sin_stub(device_type(), *this);
+  TORCH_INTERNAL_ASSERT(result.scalar_type() == output().dtype());
```
`output` is defined at this point?

hameerabbasi replied:
It's an instance method I believe.

kshitij12345 replied:
TORCH_IMPL_FUNC is 
```c++
#define TORCH_IMPL_FUNC(name) void structured_##name::impl
```

I don't see where `output().dtype()` comes from  :thinking:

hameerabbasi replied:
This becomes `at::meta::structured_sin::impl`

hameerabbasi replied:
Where `structured_sin` is a class.

kshitij12345 replied:
Oh. Makes sense.

hameerabbasi committed:
Update on "Port sin to structured."




[ghstack-poisoned]

kshitij12345 commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -346,9 +357,10 @@ Tensor& sgn_out(const Tensor& self, Tensor& result) {
-Tensor sin(const Tensor& self) { return unary_op_impl_float(self, sin_stub); }
-Tensor& sin_(Tensor& self) { return unary_op_impl_(self, at::sin_out); }
+TORCH_IMPL_FUNC(sin_out) (const Tensor& self, const Tensor& result) {
+  sin_stub(device_type(), *this);
```
https://github.com/pytorch/pytorch/blob/df837d0384398f1fac1000189d62b05e7ff0240d/aten/src/ATen/native/BinaryOps.h#L28

I think you'll have to update the stub structure similar to above.

hameerabbasi replied:
That was it, thanks!

hameerabbasi committed:
Update on "Port sin to structured."




[ghstack-poisoned]

hameerabbasi committed:
Update on "Port sin to structured."




[ghstack-poisoned]

hameerabbasi committed:
Update on "Port sin to structured."




[ghstack-poisoned]

hameerabbasi committed:
Update on "Port sin to structured."




[ghstack-poisoned]

hameerabbasi commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -661,6 +661,37 @@ static void rsqrt_kernel(TensorIterator& iter) {
   }                                                                           \
   REGISTER_DISPATCH(op##_stub, &op##_kernel)
 
+  #define IMPLEMENT_COMPLEX_STRUCTURED_KERNEL(dispatchtypes, op)                             \
```
cc @jamesr66a I may be going crazy here, but is there a difference between `TensorIterator` and `TensorIteratorBase`? It seems that I'm getting different results with each. I've just changed the former to the latter and additionally added `_STRUCTURED` to the macro.

hameerabbasi replied:
Getting these failures:

```
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_bfloat16 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=0.01, found 4084 ele...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_bool - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 2856 element(...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_complex128 - AssertionError: False is not true : Tensors failed to compare as equal!Real parts failed to compare as equal! With r...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_complex64 - AssertionError: False is not true : Tensors failed to compare as equal!Real parts failed to compare as equal! With rt...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_float32 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 4096 eleme...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_float64 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1.3e-06 and atol=1e-05, found 4096 ele...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_int16 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3930 element...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_int32 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3882 element...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_int64 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3882 element...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_int8 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3882 element(...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCPU::test_out_arg_all_dtypes_sin_cpu_uint8 - AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 3737 element...

```

jamesr66a replied:
I'm probably not the best person to ask. @ezyang @colesbury @cpuhrsch maybe?

bdhirsh replied:
I'm going to take a look at this

bdhirsh replied:
It looks like adding in a call to `TensorIteratorBase::cast_outputs()` should fix those tests. It was getting called by `unary_op_impl_float_out` originally, but isn't there after the changes in this PR.

The tests that were failing all dealt with `out` tensors that had a different dtype than the inputs. Currently when TensorIterator encounters that, it looks like it creates a temp output tensor and expects you to call `cast_outputs()` to copy the temp into the actual output tensor.

The reason you saw the failure here but we didn't see an issue porting `add_out` to structured is because binary ops all currently call `cpu_kernel()` in `Loops.h`, which handles calling `cast_outputs()`, while the unary ops in this file call `cast_outputs()` directly.

It would be nice to get this to be done more automatically by TensorIterator: it looks like Ed left a comment [here](https://github.com/pytorch/pytorch/blob/a0652c8f08f5257735f87106363258d1c14a2b52/aten/src/ATen/TensorIterator.cpp#L411) pretty much saying that, although it might take some more refactoring of TensorIterator (probably not worth doing right now). (cc @bhosmer)

I can see two reasonable ways to add it in and finish the port:
- Add a call to `iter.cast_outputs()` at the end of this `IMPLEMENT_COMPLEX_STRUCTURED_KERNEL` macro. I tried that locally and it looks like tests pass.
- Update the macro to call `cpu_kernel` / `cpu_kernel_vec` from `Loops.h`, which already has logic to call cast_outputs(). That would probably be better to cut down on code duplication, but I'm not sure how much of a drop-in code replacement that would be (i.e. if it'd break some tests).

bdhirsh replied:
and I *think* adding in the duplicate macro is reasonable to deal with any differences between them, until all of the unary ops are ported (especially given that you'll need to update the structured one)

hameerabbasi replied:
Thanks for the detailed explanation, @bdhirsh! Using `cpu_kernel_vec` fixed the test failures locally, let's see if it causes any more failures. 🤞🏼

hameerabbasi replied:
Resolving this, since some CPU tests have fully passed.

bdhirsh commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -820,6 +820,27 @@ void TensorIteratorBase::build_binary_op(const Tensor& out, const Tensor& a, con
+      .promote_integer_inputs_to_float(true));
+}
+
+void TensorIteratorBase::build_unary_op(const Tensor& out, const Tensor& a) {
```
Just to cut down on code duplication- could you update `TensorIterator::unary_op` and `unary_float_op` to call these? You can do something similar to what `TensorIterator::binary_op` currently does.

hameerabbasi replied:
Done!

hameerabbasi committed:
Update on "Port sin to structured."




[ghstack-poisoned]

hameerabbasi committed:
Update on "Port sin to structured."




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:
```diff
@@ -645,6 +645,21 @@ static void rsqrt_kernel(TensorIterator& iter) {
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(kBFloat16, iter.dtype(), #op "_vml_cpu", [&]() {\
+      cpu_kernel_vec(iter,                                                    \
+        [=](scalar_t a) __ubsan_ignore_undefined__ -> scalar_t {              \
+          return std::op(a);                                                  \
```
hmm, just comparing this code to `IMPLEMENT_ITERATOR_LAMBDA` which was there before, it looks like the current version of the unary kernels in this file use `vml` implementations ([here](https://github.com/pytorch/pytorch/blob/59ac0ff03709f6fb690918e0cd5717a0716c2eb6/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L619)) of each operator.

Judging from the headers of `Loops.h` ([here](https://github.com/pytorch/pytorch/blob/59ac0ff03709f6fb690918e0cd5717a0716c2eb6/aten/src/ATen/native/cpu/Loops.h#L25)) and `vml.h` ([here](https://github.com/pytorch/pytorch/blob/59ac0ff03709f6fb690918e0cd5717a0716c2eb6/aten/src/ATen/cpu/vml.h#L25)), I think that `cpu_kernel` might only be meant for binary ops, so we should continue using `vml` here.

@ngimel, do you happen to know if `cpu_kernel`/`cpu_kernel_vec` are only meant for binary ops (and not unary ops)? Then we should probably just keep using the existing macro here that calls `vml`, and call `cast_outputs` directly.

ngimel replied:
`cpu_kernel_vec` is used a lot in Activation.cpp for unary ops, e.g. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/Activation.cpp#L278-L287, this is clearly a unary op. 
That said, a lot of unary ops go through vml/mkl (mkl is a lot faster for special functions), and we should keep whatever implementation is used now.

bdhirsh replied:
Got it, thanks. @hameerabbasi sorry for the whiplash- we probably want to keep using the existing macro so we can keep using vml/mkl, and just tack on a call to `cast_outputs()`.

hameerabbasi replied:
Done, tests pass locally, so resolving this.

hameerabbasi committed:
Update on "Port sin to structured."




[ghstack-poisoned]

hameerabbasi commented:
The failures so far are all unrelated:

- flake8 GitHub action is failing at `test/distributed/test_c10d.py` which wasn't touched in this PR (just shows up during a merge)
- The backcompat test fails with an unrelated op:

```
Feb 25 12:54:58 Broken ops: [
Feb 25 12:54:58 	aten::linalg_multi_dot(Tensor[] tensors) -> (Tensor)
Feb 25 12:54:58 	aten::linalg_multi_dot.out(Tensor[] tensors, *, Tensor(a!) out) -> (Tensor(a!))
Feb 25 12:54:58 ]
```

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3385,16 +3385,16 @@
+  structured_delegate: sin.out
   variants: function, method
-  dispatch:
-    DefaultBackend: sin
```
@ailzhang The functional and inplace variants of `sin` used to be marked `DefaultBackend`, but now we're codegen'ing CPU/CUDA specific versions of them. That sort of technically means that if a backend had implemented `sin_out` but not `sin` or `sin_`, it would break, right?

Do you think we need to label it BC-breaking? Or is this small enough of a risk not to bother (cc @bhosmer)

hameerabbasi replied:
It might also be worth discussing if this needs to be done on a case-by-case basis as I'm hoping to port a lot more of these and then it becomes a bigger issue.

ezyang replied:
@bdhirsh We can make structured kernels automatically generate DefaultBackend kernels for the functional and inplace versions, unconditionally. This is easy to do and has no downsides.

ezyang replied:
I'm going to through up a prototype of this

ezyang replied:
(but I don't think we should block this PR on this, unless XLA is broken, which it doesn't look like it its)

bdhirsh reviewed (approved):
Like @ezyang said, since it doesn't look like this breaks XLA I'll go ahead and land this. There's a follow-up PR that'll ensure that the DefaultBackend behavior of the functional/inplace variants of the op is preserved: https://github.com/pytorch/pytorch/pull/53037/

bdhirsh commented:
@hameerabbasi do you mind rebasing?

hameerabbasi committed:
Update on "Port sin to structured."


Differential Revision: [D26732398](https://our.internmc.facebook.com/intern/diff/D26732398)

[ghstack-poisoned]

hameerabbasi commented:
> @hameerabbasi do you mind rebasing?

Done! Thanks for the quick reviews!

hameerabbasi commented:
The backcompat CI fails on an unrelated op:
```
Mar 04 14:17:18 Broken ops: [
Mar 04 14:17:18 	aten::assert_async(Tensor self) -> ()
Mar 04 14:17:18 ]
```

XLA failure unrelated also:
```
Mar 04 15:16:59 urllib.error.HTTPError: HTTP Error 403: Forbidden
```

As is the ROCm failure:
```
FAIL [0.284s]: test_stream_event_nogil (__main__.TestCuda)
```

Unfortunately I can't speak to what's broken in FB Internal.
